drivers: pinctrl: add npcm pinctrl driver

YTsea88 opened the pull request:
The changes in this PR:

- Add NPCM pinctrl driver to set pin functions and properties.
- In the device tree, set pin configuration for selection in `npcm4-pincfg-default.dtsi`, and select pinmux, pull-up/down, power stage and others like tri-state in board level `npcm4-pinctrl.dtsi`.
- Enable the NPCM pinctrl driver on the `npcm400_evb` board.

YTsea88 commented:
Will rebase once PR [#78907](https://github.com/zephyrproject-rtos/zephyr/pull/78907) is merged.

gmarull commented:
> Will rebase once PR [#78907](https://github.com/zephyrproject-rtos/zephyr/pull/78907) is merged.

Please use Draft PR to not send unnecessary noise to reviewers.

YTsea88 committed:
drivers: pinctrl : add npcm pinctrl driver

Add npcm pinctrl driver.

The pinctrl driver sets pin multiplexing with configurations in device
tree, which consist of registers and their bit values. It also sets the
peripheral pull-up/down and the device's power supply type with
configurations in the same way. Other controls, such as tri-state, are
set with raw data.

There are three tables of configurations generated from the device tree:
- npcm_pinmux_table: pin multiplexing configurations.
- npcm_periph_pupd_table: pull-up/down configurations.
- npcm_pin_lv_table: power supply configurations.

Signed-off-by: James Chiang <cpchiang1@nuvoton.com>
Signed-off-by: Alan Yang <tyang1@nuvoton.com>

gmarull reviewed (changes requested):
board level definitions seem to be in the wrong place.

YTsea88 commented:
> board level definitions seem to be in the wrong place.

Done. Move it to `zephyr/dts/arm/nuvoton/npcm/npcm4/` and rename.

gmarull commented on drivers/pinctrl/pinctrl_npcm.c:
```diff
@@ -0,0 +1,383 @@
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/dt-bindings/pinctrl/npcm-pinctrl.h>
+#include <zephyr/drivers/gpio.h>
+#include <soc.h>
```
no soc.h please, this header is meant for CMSIS integration only.

gmarull commented on drivers/pinctrl/pinctrl_npcm.c:
```diff
@@ -0,0 +1,383 @@
+#include <soc.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(pinctrl_npcm, LOG_LEVEL_ERR);
```
do not hardcode log levels

YTsea88 replied:
Fixed.

gmarull commented on drivers/pinctrl/pinctrl_npcm.c:
```diff
@@ -0,0 +1,383 @@
+		())
+
+static struct npcm_pin_lv_cfg npcm_pin_lv_table[] = {
+	DT_FOREACH_CHILD_VARGS(DT_DRV_INST(0), NPCM_PIN_LV_INIT)};
```
use `DT_INST_*` APIs, DT_DRV_INST() should be avoided.

YTsea88 replied:
Fixed.

YTsea88 commented on drivers/pinctrl/pinctrl_npcm.c:
```diff
@@ -0,0 +1,383 @@
+	    (pin->props.group >= NPCM_PINMUX_GROUP_SIZE)) {
+		LOG_ERR("No pinmux define of io%X%X", NPCM_PINCTRL_GET_PORT(pin->props.id),
+			NPCM_PINCTRL_GET_PIN(pin->props.id));
+		return;
```
Fixed. Add error code in function `pinctrl_configure_pins`.

YTsea88 commented on soc/nuvoton/npcm/common/pinctrl_soc.h:
```diff
@@ -0,0 +1,184 @@
+	uint16_t io_bias_type: 2;
+	/** Reserved field. */
+	uint16_t reserved: 14;
+} __packed;
```
I think packing is needed. Without packing, it adds one alignment byte and makes the address to access the union `cfg` in `struct npcm_pinctrl` unaligned. This would cause a compiler warning `-Waddress-of-packed-member`.

gmarull commented on dts/bindings/pinctrl/nuvoton,npcm-pinctrl.yaml:
```diff
@@ -0,0 +1,155 @@
+# Copyright (c) 2024 Nuvoton Technology Corporation.
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
```
wrong commit, needs to be part of the 1st

YTsea88 replied:
Fixed.

gmarull commented on dts/bindings/pinctrl/nuvoton,npcm-pinctrl.yaml:
```diff
@@ -0,0 +1,155 @@
+        description: |
+          The pin configuration group property.
+          Configurations of the device control such as tri-state, IO type, and so on.
+      lvol:
```
please prefix all vendor specific properties with `nuvoton,`

YTsea88 replied:
Fixed.

YTsea88 committed:
dts: arm: nuvoton: add npcm pinctrl instances

Add npcm pinctrl instances and device tree bindings for npcm pinctrl.
Set the pinmux alternate functions, which can be selected in the pinctrl
child node.

Signed-off-by: Alan Yang <tyang1@nuvoton.com>

YTsea88 commented on dts/arm/nuvoton/npcm/npcm4/npcm4-pincfg-default.dtsi:
```diff
@@ -0,0 +1,1822 @@
+				NPCM_PIN_REG_NULL>;
+		nuvoton,alt = <0 0x10 0x8 NPCM_PIN_CFG_NULL NPCM_PIN_CFG_NULL NPCM_PIN_CFG_NULL>;
+		nuvoton,pupd-reg = <NPCM_PINCTRL_DEVPU2_1_SMB8A_PUE>;
+		nuvoton,pupd-type = <NPCM_BIAS_TYPE_PULL_UP>;
```
Sure, change to standard properties and remove redundant types.

gmarull commented on dts/bindings/pinctrl/nuvoton,npcm-pinctrl.yaml:
```diff
@@ -0,0 +1,164 @@
+    i2c1a_scl_sda_gpio60_61: periph-i2c1a {
+      scl {
+        pinmux = <GPIO60__RSTOUT2__SCL1A 2>;  /* Set the pinmux to SCL1A */
+        lvol = <0>;                           /* Set the voltage level to 3.3V */
```
this looks unreadable, description should be clear from a DT point of view: standard `power-source` may be used, or, `vendor,voltage-level-mv = <3300>`.

YTsea88 replied:
Fixed. Also rename the `pupd-*`, `lv-*` and `dev-ctl` properties to make it clear in DT.

gmarull commented on dts/bindings/pinctrl/nuvoton,npcm-pinctrl.yaml:
```diff
@@ -0,0 +1,164 @@
+        The peripheral pull-up/down definition property.
+        The peripheral pull-up/down register and its bit. There are a total of 2 registers if
+        property pupd-config-extend is set.
+    bias-pull-up:
+      type: boolean
+      description: |
+        The peripheral pull-up/down definition property.
+        The default peripheral pull-up in hardware of supported alt functions.
+    bias-pull-down:
+      type: boolean
+      description: |
+        The peripheral pull-up/down definition property.
+        The default peripheral pull-down in hardware of supported alt functions.
```
please, do not redefine standard properties.

YTsea88 replied:
Removed.

gmarull reviewed (changes requested):
still has some issues

gmarull commented on dts/bindings/pinctrl/nuvoton,npcm-pinctrl.yaml:
```diff
@@ -0,0 +1,154 @@
+        type: array
+        required: true
+        description: |
+          The pin configuration group property.
+          The alternate pin group index and selection of alternate function.
+      nuvoton,device-control:
+        type: array
+        description: |
+          The pin configuration group property.
+          Configurations of the device control such as tri-state, IO type, and so on.
+      nuvoton,voltage-level-mv:
+        type: int
+        description: |
+          The pin configuration group property.
```
descriptions are misleading, please provide a more concise description/examples in the binding docs.

YTsea88 replied:
Done.

gmarull commented on include/zephyr/dt-bindings/pinctrl/npcm-pinctrl.h:
```diff
@@ -0,0 +1,536 @@
+#define NPCM_IO_VOLT_LV_3V3 0
+#define NPCM_IO_VOLT_LV_1V8 1
+
+/* Supported IO bias type in NPCM series */
+#define NPCM_BIAS_TYPE_NONE      0
+#define NPCM_BIAS_TYPE_PULL_UP   (1 << 0)
+#define NPCM_BIAS_TYPE_PULL_DOWN (1 << 1)
```
how is this used in DT? this file must **only** contain bits used in DT.

YTsea88 replied:
Fix and correct description.

gmarull commented on dts/bindings/pinctrl/nuvoton,npcm-pinctrl.yaml:
```diff
@@ -0,0 +1,191 @@
+        - bias-pull-up
+
+  properties:
+    nuvoton,pinmux-group-def:
+      type: boolean
+      description: Indicates that this node is a pinmux group definition.
```
can you please explain this? is this related to hw description in any way?

YTsea88 replied:
This is the hardware definition to the pin. Basically, to set up config of a pin, it needs this pin **definition** node in `npcm4-pincfg-default.dtsi` to describe the hardware with all the pinmux and supported voltage levels, etc. Then select the pin function and voltage, ... in the pin **configuration** node in `npcm4-pinctrl.dtsi`.

Property renamed to `nuvoton,pin-hw-def` and rewrite the descriptions for better understanding.

gmarull commented on dts/bindings/pinctrl/nuvoton,npcm-pinctrl.yaml:
```diff
@@ -0,0 +1,191 @@
+      description: |
+        The pinmux group definition property.
+        The pin's index. The name of this property follows the group's order.
+    nuvoton,low-voltage-reg:
+      type: array
+      description: |
+        The low-voltage level definition property.
+        The low-voltage level register and its bit.
```
needs better explanation

YTsea88 replied:
Fixed.

gmarull commented on dts/bindings/pinctrl/nuvoton,npcm-pinctrl.yaml:
```diff
@@ -0,0 +1,191 @@
+      description: |
+        The low-voltage level definition property.
+        The low-voltage level register and its bit.
+    nuvoton,low-voltage-def:
+      type: array
+      description: |
+        The low-voltage level definition property.
+        The default voltage level in hardware of supported alt functions. Set with either
+        NPCM_IO_VOLT_LV_1V8 or NPCM_IO_VOLT_LV_3V3.
```
why not `low-voltage-mv = <1800>/<3300>`?

YTsea88 replied:
It is stored as a Boolean to save space. I think it can be represented in mV for better readability.

YTsea88 commented:
Also remove the duplicated EMC control register definition.

fabiobaltieri commented on drivers/pinctrl/pinctrl_npcm.c:
```diff
@@ -0,0 +1,399 @@
+	 DT_PROP_BY_IDX(node_id, prop, 4), DT_PROP_BY_IDX(node_id, prop, 5)}
+
+#define NPCM_PINMUX_INIT(node_id, prop)                                                            \
+	COND_CODE_1( \
```
I think you can use `IF_ENABLED` here and drop the empty "else" case here and below

YTsea88 replied:
Sure, all changed to `IF_ENABLED`.

fabiobaltieri commented on drivers/pinctrl/pinctrl_npcm.c:
```diff
@@ -0,0 +1,399 @@
+#define NPCM_PIN_PERIPH_PUPD_INIT(node_id, prop)                                                   \
+	COND_CODE_1( \
+		DT_PROP(node_id, nuvoton_pin_hw_def), \
+		(COND_CODE_1( \
```
same here and above, should make this a bit easier to read

YTsea88 replied:
Fixed.

fabiobaltieri commented on drivers/pinctrl/pinctrl_npcm.c:
```diff
@@ -0,0 +1,399 @@
+#define NPCM_PIN_LV_INIT(node_id, prop)                                                            \
+	COND_CODE_1( \
+		DT_PROP(node_id, nuvoton_pin_hw_def), \
+		(COND_CODE_1( \
```
here too

YTsea88 replied:
Fixed.

fabiobaltieri commented on drivers/pinctrl/pinctrl_npcm.c:
```diff
@@ -0,0 +1,399 @@
+#define NPCM_PINCTRL_GET_PIN(pin)  (((pin) >> 0) & NPCM_PINCTRL_PIN_MASK)
+
+/* Voltage level constants */
+#define NPCM_VOLT_LV_3V3_CONST 3300
+#define NPCM_VOLT_LV_1V8_CONST 1800
```
drop the `CONST` suffix, it's fairly clear that it's a constant since it's uppercase

YTsea88 replied:
Fixed.

fabiobaltieri commented on drivers/pinctrl/pinctrl_npcm.c:
```diff
@@ -0,0 +1,399 @@
+	const uintptr_t scfg_base = npcm_pinctrl_cfg.base_scfg;
+	uint8_t cfg_value;
+	uint16_t reg_value;
+	int i;
```
could this be a uint8_t? save a bit of stack

YTsea88 replied:
Sure, change to `uint8_t`.

fabiobaltieri commented on drivers/pinctrl/pinctrl_npcm.c:
```diff
@@ -0,0 +1,399 @@
+
+	/* Get pupd register and type */
+	pupd_type = npcm_periph_pupd_table[io].pupd_type;
+	if (reg_idx == 0) {
```
are there constants for 0 and 1 here?

YTsea88 replied:
Add const `NPCM_PERIPH_PUPD_TABLE_REG0_IDX` and `NPCM_PERIPH_PUPD_TABLE_REG1_IDX`.

fabiobaltieri commented on drivers/pinctrl/pinctrl_npcm.c:
```diff
@@ -0,0 +1,399 @@
+static int npcm_periph_pupd_configure(const struct npcm_pinctrl *pin)
+{
+	const struct npcm_periph_pupd *pupd = &pin->cfg.pupd;
+	int sup_idx = 0;
+	int i;
+	int j;
```
can these use a smaller type?

YTsea88 replied:
Change to `uint8_t` and add build assertion on table size checking.

fabiobaltieri commented on soc/nuvoton/npcm/common/pinctrl_soc.h:
```diff
@@ -0,0 +1,195 @@
+
+#define Z_PINCTRL_STATE_PIN_INIT_EXT(node_id, prop)                                                \
+	Z_PINCTRL_NPCM_PERIPH_PINMUX_INIT(node_id)                                                 \
+	COND_CODE_1( \
```
same comment as above for IF_ENABLED

YTsea88 replied:
Fixed.

fabiobaltieri commented on soc/nuvoton/npcm/common/pinctrl_soc.h:
```diff
@@ -0,0 +1,195 @@
+	COND_CODE_1( \
+		DT_NODE_HAS_PROP(node_id, pinmux), \
+		(Z_PINCTRL_STATE_PIN_INIT_EXT(node_id, prop)), \
+		())
```
IF_ENABLED again

YTsea88 replied:
Fixed.

gmarull commented on dts/bindings/pinctrl/nuvoton,npcm-pinctrl.yaml:
```diff
@@ -0,0 +1,193 @@
+        - bias-pull-up
+
+  properties:
+    nuvoton,pin-hw-def:
+      type: boolean
+      description: Indicates that this node is a pin's hardware definition that includes pinmux
+        group, low-voltage level, and peripheral pull-up/down properties.
```
how is this related to hw description? what do other nodes describe then?

The whole DT here is confusing, needs a better explanation, maybe some high-level document and how decisions have been made. By reading properties descriptions I even get more confused.

YTsea88 replied:
I think it can be simplified with an example of drivers that actually control the pins.
Take the i2c driver as an example:

- Set GPIO17 to SDA
- Set GPIO16 to SCL


By default, each pin's alternate function settings, including the registers and bits, are defined in the DTS node with the property `nuvoton,pin-hw-def`. These are the definitions of the pin hardware. The figure below shows the pin hardware definition for GPIO17 and GPIO16.
The tables of all the pin's hardware definitions of pinmux, voltage level, and pull-up/down are generated at build time with a unique ID macro named in format of _GPIOn__FUNC1__FUNC2__FUNC3_.
![image](https://github.com/user-attachments/assets/6554b863-f397-4094-8338-61ac3099dfa4)


Next, the configurations of the I2C pins are configured in the DTS node _i2c6b_scl_sda_gpio16_17_ in the figure below.
![image](https://github.com/user-attachments/assets/aa96bf0f-8587-4f5a-b215-691410a79074)
In this example, in `zephyr/drivers/*/*_i2c.c`:
On initialization, a `pinctrl_soc_pin_t` array is defined
(1) Pin ID _GPIO17__CLKOUT__RIA__SDA6B_ idx 3: pinmux SDA
(2) Pin ID _GPIO17__CLKOUT__RIA__SDA6B_ idx 3: voltage level 3.3V
(3) Pin ID _GPIO17__CLKOUT__RIA__SDA6B_ idx 3: pull-up
(4) Pin ID _GPIO16__DCDA__SCL6B_ idx 2: pinmux SCL

Then control the pins with pinctrl API: `pinctrl_apply_state()` -> … -> vendor defined `pinctrl_configure_pins()`.
In `pinctrl_configure_pins()`, find the target definition of the config by ID. For example, in array (1), the ID is _GPIO17__CLKOUT__RIA__SDA6B_. And search for the index of this ID _GPIO17__CLKOUT__RIA__SDA6B_ in `npcm_pinmux_table`, get the corresponding settings of index 3 (0x4 in this case), and set the regs and bit.

YTsea88 replied:
@gmarull Is the explanation clear and easy to understand?
